### PR TITLE
WIP: migrate to pnpm

### PR DESCRIPTION
**Pull request by 6543**
at the moment `pnpm webpack` fails, this need investigation

in general, it would speedup frontend build/lint/etc... a lot

https://pnpm.io/benchmarks

TODOs:
- [ ] make build (webpack) work
- [ ] make sure we do have a win with this
- [ ] update docs
- [ ] does have packagers an issue with it?

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -240,9 +240,9 @@ node-check:
 	$(eval MIN_NODE_VERSION_STR := $(shell grep -Eo '"node":.*[0-9.]+"' package.json | sed -n 's/.*[^0-9.]\([0-9.]*\)"/\1/p'))
 	$(eval MIN_NODE_VERSION := $(shell printf "%03d%03d%03d" $(shell echo '$(MIN_NODE_VERSION_STR)' | tr '.' ' ')))
 	$(eval NODE_VERSION := $(shell printf "%03d%03d%03d" $(shell node -v | cut -c2- | tr '.' ' ');))
-	$(eval NPM_MISSING := $(shell hash npm > /dev/null 2>&1 || echo 1))
-	@if [ "$(NODE_VERSION)" -lt "$(MIN_NODE_VERSION)" -o "$(NPM_MISSING)" = "1" ]; then \
-		echo "Gitea requires Node.js $(MIN_NODE_VERSION_STR) or greater and npm to build. You can get it at https://nodejs.org/en/download/"; \
+	$(eval PNPM_MISSING := $(shell hash pnpm > /dev/null 2>&1 || echo 1))
+	@if [ "$(NODE_VERSION)" -lt "$(MIN_NODE_VERSION)" -o "$(PNPM_MISSING)" = "1" ]; then \
+		echo "Gitea requires Node.js $(MIN_NODE_VERSION_STR) or greater and pnpm to build. You can get it at https://nodejs.org/en/download/"; \
```
Comment (6543) [inline Makefile:245]: we might need to update that hint

**Comment by silverwind** [comment]
I'm -1 on this, I see no benefit.

- First-time module installation [will be faster](https://pnpm.io/benchmarks) but for subsequent executions, performance is about equal.
- npm is ubiqutious while pnpm (and yarn) is esotheric.
- It would complicate from-source build for all users as they now need to install this additional dependency (npm always comes with node).

**Comment by lafriks** [comment]
Problem with npm (and yarn) is not so much about speed but disk usage. If you are working on single project it's not big a deal but if you have many projects with each having 1GB node_modules directories this add up quite much...

**Comment by mohe2015** [comment]
Is it not possible to make this pnpm compatible and also note in the README but not switch to it by default?

**Comment by silverwind** [comment]
> Problem with npm (and yarn) is not so much about speed but disk usage. If you are working on single project it's not big a deal but if you have many projects with each having 1GB node_modules directories this add up quite much...

Is disk space really a problem in the current year? A modern development machine probably has 1TB or more, 1GB is just 0.1% of that disk space. Also, on my machine, gitea's node_modules is just 350MB currently.

> Is it not possible to make this pnpm compatible and also note in the README but not switch to it by default?

Not without us having the burden of managing two lockfiles. Or does pnpm include compatibilty for `package-lock.json`? If so, it would already be compatible.

**Comment by wxiaoguang** [comment]
Looks like most people agree with staying with npm ?

**Comment by wxiaoguang** [comment]
Let's close it ~~~

**Pull request closed without merging**